rosidl_generator_py/resource/_msg_support.c.template

```diff
@@ -42,73 +55,161 @@ full_classname = "@(spec.base_type.pkg_name).@(subfolder)._@(module_name).@(spec
   snprintf(full_classname_dest, @(len(full_classname) + 1), "%s.%s", class_name, module_name);
 
   assert(strncmp(
-    "@(spec.base_type.pkg_name).@(subfolder)._@(module_name).@(spec.base_type.type)",
-    full_classname_dest, @(len(full_classname))));
+      "@(spec.base_type.pkg_name).@(subfolder)._@(module_name).@(spec.base_type.type)",
+      full_classname_dest, @(len(full_classname))));
 
 @[for field in spec.fields]@
   PyObject * py@(field.name) = PyObject_GetAttrString(_pymsg, "@(field.name)");
-@[if not field.type.is_primitive_type()]@
+@[  if not field.type.is_primitive_type()]@
 @{
 nested_type = '%s__%s__%s' % (field.type.pkg_name, 'msg', field.type.type)
 }@
-@[if field.type.is_array]@
-  // TODO(mikael) implement nested array
-  assert(PySequence_Check(py@(field.name)));
-@[else]@
   PyObject * py@(field.name)_msg_module = PyImport_ImportModule("@(field.type.pkg_name).msg._@convert_camel_case_to_lower_case_underscore(field.type.type)");
   PyObject * py@(field.name)_convert_from_py = PyObject_GetAttrString(py@(field.name)_msg_module, "_convert_from_py");
   typedef PyObject *(* convert_from_py_signature)(void *);
   convert_from_py_signature convert_from_py_@(field.name) = (convert_from_py_signature)PyCapsule_GetPointer(py@(field.name)_convert_from_py, NULL);
+@[    if field.type.is_array]@
+  assert(PySequence_Check(py@(field.name)));
+  PyObject * seq@(field.name) = PySequence_Fast(py@(field.name), "expected a sequence");
+  @(nested_type) * item@(field.name);
+@[      if field.type.array_size is None]@
+  size_t size@(field.name) = PySequence_Size(py@(field.name));
+@[      else]@
+  size_t size@(field.name) = @(field.type.array_size);
+@[      end if]@
+  @(nested_type) * tmparray@(field.name) = (@(nested_type) *)malloc(sizeof(@(nested_type)) * size@(field.name));
+  size_t idx@(field.name);
+  for (idx@(field.name) = 0; idx@(field.name) < size@(field.name); idx@(field.name)++) {
+    item@(field.name) = (@(nested_type) *) convert_from_py_@(field.name)(
+      PySequence_Fast_GET_ITEM(seq@(field.name), idx@(field.name)));
+    tmparray@(field.name)[idx@(field.name)] = *item@(field.name);
+  }
+@[      if field.type.array_size is None]@
+  if (!@(nested_type)__Array__init(&(ros_message->@(field.name)), size@(field.name))) {
+    assert(false && "unable to create array ros_message");
```

Comment:
This probably shouldn't use an assert, since it will not be checked in Release mode.